File: 2025/day11.go
package main

import (
	"fmt"
	"slices"
	"strings"

	"aoc/tools"
	_ "embed"
)

var DEBUG bool = false
var TEST bool = false
var test = `
aaa: you hhh
you: bbb ccc
bbb: ddd eee
ccc: ddd eee fff
ddd: ggg
eee: out
fff: out
ggg: out
hhh: ccc fff iii
iii: out`[1:]

var test2 = `
svr: aaa bbb
aaa: fft
fft: ccc
bbb: tty
tty: ccc
ccc: ddd eee
ddd: hub
hub: fff
eee: dac
dac: fff
fff: ggg hhh
ggg: out
hhh: out`[1:]

//go:embed day11.txt
var live string

type network = map[string][]string

func parse(data []string) network {
	connx := make(network)
	for _, line := range data {
		parts := strings.Split(line, " ")
		name := parts[0][:len(parts[0])-1]
		connx[name] = parts[1:]
	}
	return connx
}

var cache map[string]int

func paths(connx network, start string, end string) int {
	val, ok := cache[start+end]
	if ok {
		return val
	}

	sum := 0
	if slices.Contains(connx[start], end) {
		sum = 1
	}
	for _, n := range connx[start] {
		sum += paths(connx, n, end)
	}
	cache[start+end] = sum
	return sum
}

func part1(data []string) int {
	connx := parse(data)
	return paths(connx, "you", "out")
}

func part2(data []string) int {
	connx := parse(data)
	return 0 +
		paths(connx, "svr", "dac")*paths(connx, "dac", "fft")*paths(connx, "fft", "out") +
		paths(connx, "svr", "fft")*paths(connx, "fft", "dac")*paths(connx, "dac", "out")
}

func main() {
	var input string
	TEST, DEBUG, input = tools.Setup(test, live)
	data := strings.Split(input, "\n")

	cache = make(map[string]int)
	fmt.Println("Part 1:", part1(data))

	if TEST {
		data = strings.Split(test2, "\n")
	}
	cache = make(map[string]int)
	fmt.Println("Part 2:", part2(data))
}

File: 2025/day11.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <iterator>
#include <cstdint>
#include <cstring>
#include <cmath>
#include <vector>
#include <map>
#include <set>
#include <queue>
#include <algorithm>
#include <numeric>

#include "utils.h"

using namespace std;

const string test(
    "aaa: you hhh\n"
    "you: bbb ccc\n"
    "bbb: ddd eee\n"
    "ccc: ddd eee fff\n"
    "ddd: ggg\n"
    "eee: out\n"
    "fff: out\n"
    "ggg: out\n"
    "hhh: ccc fff iii\n"
    "iii: out"
);

const string test2(
    "svr: aaa bbb\n"
    "aaa: fft\n"
    "fft: ccc\n"
    "bbb: tty\n"
    "tty: ccc\n"
    "ccc: ddd eee\n"
    "ddd: hub\n"
    "hub: fff\n"
    "eee: dac\n"
    "dac: fff\n"
    "fff: ggg hhh\n"
    "ggg: out\n"
    "hhh: out"
);

#define DAY "day11"

bool DEBUG = false;
bool TEST = false;

typedef map<string,list<string>> network_t;

network_t parse(StringVector & data )
{
    network_t connx;
    for( auto & line : data )
    {
        auto parts = split_list(line, " ");
        string name = parts.front();
        name = name.substr(0, name.size()-1);
        parts.pop_front();
        connx[name] = parts;
    }
	return connx;
}

map<string,int64_t> cache;

int64_t paths( network_t & connx, const string & start, const string & end )
{
    auto pr = cache.find(start+end);
    if( pr != cache.end() )
        return pr->second;

    int64_t sum = 0;
    if( find(connx[start].begin(), connx[start].end(), end) != connx[start].end() )
        sum = 1;

    for( auto n : connx[start] )
        sum += paths( connx, n, end );

	cache[start+end] = sum;
	return sum;
}


int64_t part1 ( StringVector & data )
{
    network_t connx = parse(data);
    return paths(connx, "you", "out");
}

int64_t part2 ( StringVector & data )
{
    network_t connx = parse(data);
    return paths(connx, "svr", "fft") * paths(connx, "fft", "dac") * paths(connx, "dac", "out") +
           paths(connx, "svr", "dac") * paths(connx, "dac", "fft") * paths(connx, "fft", "out");
}

int main( int argc, char ** argv )
{
    string name = *argv;
    while( *++argv )
    {
        string arg(*argv);
        if( arg == "debug")
            DEBUG = true;
        else if( arg =="test")
            TEST = true;
    }

    string input = TEST ? test : file_contents(DAY".txt");
    auto data = split(input);

    cout << "Part 1: " << part1(data) << "\n";
    cache.clear();
    if( TEST )
        data = split(test2);
    cout << "Part 2: " << part2(data) << "\n";
}
